File: apps/desktop/src/lib/pdf/papers.ts
/** The words of the papers in a space, and searching them.
 *
 *  A paper is not a note: its words are inside a PDF, and getting at them is
 *  pdf.js, a worker and a page at a time. So the space search cannot walk them the
 *  way it walks notes - neither the crate nor the browser's own walk has anything
 *  to read - and what nib does instead is take a page's words down the first time
 *  anything reads them and keep them:
 *
 *  In memory for the sitting, which is what the viewer fills as a reader scrolls,
 *  since it reads a page's text for its own find bar anyway. And on disk, keyed by
 *  the file's hash, which is what makes a paper opened last week answer a search
 *  today; see text-cache.ts, which owns the store and the bounds, and extract.ts,
 *  which reads the papers nobody has opened at all in idle time.
 *
 *  What was refused, and why. Reading every PDF in a space when a search runs
 *  would make the first search of a space with twenty papers a minute long, which
 *  is not a search. Nothing here ever waits for a PDF to be taken apart: a search
 *  answers from what is in memory, and everything that puts something there
 *  happens in idle time or because somebody opened the paper.
 *
 *  Matching itself is the shared matcher, so a query means the same thing in a
 *  paper as in a note. Only the rows are built here: a page has no lines, so the
 *  words around the match are what a row shows. */

import { type Hit, Matcher } from '../search/match'
import type { Query } from '../search/query'
import { warm } from '../search/warm.svelte'
import { afterQuiet } from '../timing'
import { forgetPaperText, keepPaperText, paperFiles, paperText, paperTextOf } from './text-cache'

/** One paper, as its words stand in this window. */
interface Paper {
  /** The hash of the file the words came out of, where it is known: the viewer
   *  and the idle pass both have the bytes in hand and both say. Empty for a
   *  paper whose words arrived before anybody said, which is nothing to write
   *  down - a record nobody can check the file against is a record that would
   *  outlive the file. */
  hash: string
  modified: number
  pages: Map<number, string>
  characters: number
  /** Whether the pages have changed since they were last written down. */
  unwritten: boolean
}

/** Every paper this window has words for, by the paper's path. */
const held = new Map<string, Paper>()

/** How much paper text is kept in memory at once.
 *
 *  Less than the store holds, because this is a page and that is a disk: what
 *  comes back after a restart is read smallest first, so as many papers as
 *  possible answer rather than one book filling the whole of it. */
const IN_MEMORY = 4_000_000

/** How long after the last page was read its paper is written down. Long enough
 *  that scrolling through twenty pages is one write, short enough that closing the
 *  tab a moment later has still kept them. */
const SETTLE = 1500

/** How much of a page a row shows, and how much of it comes before the match. A
 *  page is one long run of words, so unlike a note's line there is no natural end
 *  to cut at: the row is a window, and the match sits inside it rather than at the
 *  start of it. */
const WINDOW = 180
const BEFORE = 60

/** The papers written down once the reader has stopped turning pages; see
 *  `pageRead` and `writePapers`. */
const writing = afterQuiet(() => void writePapers(), SETTLE)

/** Which space has had what was taken down in an earlier sitting read back, so
 *  that a search asks for it once rather than per keystroke. */
let hydrated: { root: string; done: Promise<void> } | null = null

/** Which file a paper's words came out of.
 *
 *  Said by whoever has the bytes: the viewer when it opens a paper, and the idle
 *  pass when it reads one nobody has opened. Until something says, a page's words
 *  are held for the sitting and not written down, because a record that cannot be
 *  held against the file is one that would answer for a file that has changed. */
export function paperOpened(path: string, hash: string, modified: number): void {
  const paper = paperAt(path)
  if (paper.hash === hash && paper.modified === modified) return

  // A different file at the same path: what was held was that file's words.
  if (paper.hash && paper.hash !== hash) {
    paper.pages.clear()
    paper.characters = 0
    void forgetPaperText(path)
  }

  paper.hash = hash
  paper.modified = modified
}

function paperAt(path: string): Paper {
  const found = held.get(path)
  if (found) return found

  const made: Paper = { hash: '', modified: 0, pages: new Map(), characters: 0, unwritten: false }
  held.set(path, made)
  return made
}

/** A page's words, as the viewer read them.
 *
 *  The runs are joined with a space rather than with nothing: `textOfRuns` joins
 *  them tight because the find bar counts characters against what is painted, and
 *  nothing here is painted. A space is what keeps two runs from making one word
 *  that neither of them says. */
export function paperRead(path: string, page: number, runs: readonly string[]): void {
  const text = runs.join(' ')
  const paper = paperAt(path)
  const was = paper.pages.get(page)
  if (was === text) return

  paper.pages.set(page, text)
  paper.characters += text.length - (was?.length ?? 0)
  paper.unwritten = true

  hold()
  told()
  // Written down once the reader has stopped turning pages, and only for a paper
  // whose file has been named; see `paperOpened`.
  writing()
}

/** Writes down every paper whose pages have changed since the last time.
 *
 *  Awaited by the idle pass, which reads one paper and writes it before it starts
 *  the next, and by the tests. The viewer's own pages are written on a timer. */
export async function writePapers(): Promise<void> {
  writing.cancel()

  for (const [path, paper] of held) {
    if (!paper.unwritten || !paper.hash || !paper.pages.size) continue

    paper.unwritten = false
    await keepPaperText(paperTextOf(path, paper.hash, paper.modified, paper.pages))
  }
}

/** Reads back what was taken down in an earlier sitting, once per space.
 *
 *  Awaited by the space search before it asks the papers anything, and asked for
 *  in idle time at the search stage of the launch, so the first question about a
 *  space has the answer already. Smallest first: a shelf of papers answers rather
 *  than one book filling everything there is.
 *
 *  A record is read with what the listing says about the file, so a paper written
 *  since it was taken down is dropped rather than answered with; see
 *  text-cache.ts. */
export async function papersFor(root: string, modified?: Map<string, number>): Promise<void> {
  if (hydrated?.root === root) return hydrated.done

  const done = (async () => {
    const inside = root.endsWith('/') ? root : `${root}/`

    for (const file of await paperFiles()) {
      if (!file.path.startsWith(inside)) continue
      if (characters() + file.size > IN_MEMORY) break
      if (held.get(file.path)?.pages.size) continue

      const record = await paperText(file.path, modified?.get(file.path))
      if (!record) continue

      const paper = paperAt(file.path)
      paper.hash = record.hash
      paper.modified = record.modified
      for (const [page, text] of record.pages) paper.pages.set(page, text)
      paper.characters = record.characters
      paper.unwritten = false
    }

    told()
  })()

  hydrated = { root, done }
  return done
}

/** Forgets a paper, for one that has been renamed, moved or deleted. Its words go
 *  from the store as well: they were that path's words, and the path has gone. */
export function paperGone(path: string): void {
  held.delete(path)
  told()
  void forgetPaperText(path)
}

/** A paper, or a folder of them, that has moved.
 *
 *  The words follow the file, because they are the file's: what was read out of
 *  those bytes is still what is in them wherever they sit. The record moves with
 *  them, the way a PDF's highlights follow it; see `move_highlights` in paths.rs. */
export function paperMoved(from: string, to: string): void {
  const moving = [...held.keys()].filter((path) => path === from || path.startsWith(`${from}/`))

  for (const path of moving) {
    const paper = held.get(path)
    if (!paper) continue

    held.delete(path)
    held.set(to + path.slice(from.length), { ...paper, unwritten: true })
    void forgetPaperText(path)
  }

  if (!moving.length) return

  told()
  void writePapers()
}

/** Says what is held, so that the panel's diagnostics line says both halves of
 *  what a search reads; see search/warmth.ts. */
function told(): void {
  warm.holdingPapers(papersHeld())
}

/** What is held, for the panel's diagnostics and for a test. */
export function papersHeld(): { papers: number; characters: number; cap: number } {
  return { papers: papersRead(), characters: characters(), cap: IN_MEMORY }
}

/** How many papers have words worth searching. For the drive, and for a test. */
export function papersRead(): number {
  let read = 0
  for (const paper of held.values()) {
    if (paper.pages.size) read += 1
  }

  return read
}

function characters(): number {
  let sum = 0
  for (const paper of held.values()) sum += paper.characters
  return sum
}

/** Memory held to `IN_MEMORY`, the largest paper first: a paper let go of is one
 *  the store still has, so it comes back the next time the space is opened. */
function hold(): void {
  if (characters() <= IN_MEMORY) return

  const biggest = [...held.entries()].sort((one, other) => other[1].characters - one[1].characters)
  for (const [path, paper] of biggest) {
    if (characters() <= IN_MEMORY) return
    // Never the paper whose pages are still unwritten: those words are only here.
    if (paper.unwritten) continue

    held.delete(path)
  }
}

/** The words around one match, and where the match sits in them. */
function windowOf(body: string, from: number): { text: string; at: number } {
  const start = Math.max(0, from - BEFORE)
  const cut = body.slice(start, start + WINDOW)
  const lead = cut.length - cut.trimStart().length
  const text = cut.trim()

  return { text, at: from - start - lead }
}

/** Whether a paper is one the space leaves out: the path itself, or something
 *  inside a folder that is. The same reading the notes get; a list of papers is a
 *  handful, so the list is walked here rather than the path's own ancestors. */
function leftOut(relative: string, excluded: readonly string[]): boolean {
  return excluded.some((one) => relative === one || relative.startsWith(`${one}/`))
}

/** What the papers of one space answer, as the rows the panel draws.
 *
 *  A row carries its page rather than a line, which is what says it is a paper: the
 *  panel opens it there, and a paper has no lines to go to. */
export function searchPapers(
  root: string,
  query: Query,
  limit: number,
  excluded: readonly string[] = [],
): Hit[] {
  if (!held.size || limit <= 0) return []

  const matcher = new Matcher(query)
  const out: Hit[] = []
  const inside = root.endsWith('/') ? root : `${root}/`

  for (const [path, paper] of held) {
    if (out.length >= limit) break
    if (!path.startsWith(inside)) continue

    const relative = path.slice(inside.length)
    if (leftOut(relative, excluded)) continue

    const name = relative.split('/').pop() ?? relative

    for (const [page, body] of [...paper.pages].sort((one, other) => one[0] - other[0])) {
      if (out.length >= limit) break
      if (!body.trim()) continue

      const spans = matcher.spans({ path, relative, name, body })
      if (!spans) continue

      // A page answers once: it is one place in the paper, and a row per word
      // found on it would be a page of rows about one page.
      const first = spans[0]
      const { text, at } = first
        ? windowOf(body, first.from)
        : { text: body.slice(0, WINDOW).trim(), at: -1 }

      const length = first ? first.to - first.from : 0
      out.push({
        path,
        name,
        page,
        // A paper has no lines. Nothing reads this for a paper - the page is what
        // opens one - and zero is what a row with no line says.
        line: 0,
        text,
        ranges: at >= 0 && length > 0 ? [{ from: at, to: at + length }] : [],
      })
    }
  }

  return out
}

/** For the tests: nothing read, as at launch. What is in the store stays there,
 *  which is what the store is for. */
export function forgetPapers(): void {
  held.clear()
  hydrated = null
  writing.cancel()
}

File: apps/desktop/src-tauri/src/papers.rs
//! The words taken out of a PDF, kept so that they are only taken out once.
//!
//! A PDF's words are not text on disk: getting at them is the window's own pdf.js,
//! a worker and a page at a time. So the window takes them down as they are read
//! and hands them here, and a paper somebody opened last week answers a search
//! today without being opened again.
//!
//! They live in the app's own config folder rather than beside the PDF, which is
//! the rule the version history follows for the same reason: a PDF's folder belongs
//! to the person who put the PDF there, and a cache is the app's business. Each
//! paper gets a folder named after a hash of its path - `folder_key` in `paths`,
//! the same one history uses - holding the words and the path they came from, so
//! that the store can be listed by paper without every record in it being read.
//!
//! What is in a record belongs to the window, which is the only thing that reads
//! it: here it is text, judged only by its size and by where it may be written. The
//! window keys it by the hash of the file's bytes and drops it when that changes;
//! see pdf/text-cache.ts.

use serde::Serialize;
use std::fs;
use std::path::{Path, PathBuf};
use tauri::{AppHandle, Manager};

use crate::paths::{cannot, folder_key, is_pdf, write_atomically};

/// How much of one paper's words may be kept. A thousand pages of a dense book,
/// and a ceiling so that a window asking a cache to hold more than a cache should
/// is refused rather than obeyed. The window holds itself to the same number; this
/// is the one that protects the disk.
const LIMIT: u64 = 4 * 1024 * 1024;

/// What the store holds for one paper: which paper, and how much.
#[derive(Serialize)]
pub struct PaperFile {
    /// The PDF's path, as the window spells it.
    path: String,
    /// How many bytes its words come to.
    size: u64,
}

/// The folder every paper's words live under.
fn papers_dir(app: &AppHandle) -> Result<PathBuf, String> {
    Ok(app
        .path()
        .app_config_dir()
        .map_err(|error| format!("could not find the settings folder: {error}"))?
        .join("papers"))
}

/// One paper's folder. The path is judged first: a store of what is inside PDFs
/// holds nothing else, so a window that names something that is not one is refused
/// rather than answered.
fn paper_dir(app: &AppHandle, pdf: &str) -> Result<PathBuf, String> {
    if !is_pdf(Path::new(pdf)) {
        return Err(format!("{pdf} is not a PDF"));
    }

    Ok(papers_dir(app)?.join(folder_key(pdf)))
}

/// Where the words are, and where the paper they came from is named. Two files
/// rather than one, so that listing the store reads a path per paper instead of
/// megabytes of words; `origin.txt` is the same trick history.rs plays.
fn pages_file(dir: &Path) -> PathBuf {
    dir.join("pages.json")
}

fn origin_file(dir: &Path) -> PathBuf {
    dir.join("origin.txt")
}

/// What was taken down for one paper, or nothing at all. A record larger than a
/// record may be is nothing rather than an error: the paper is read again, and the
/// next write is what clears it.
fn read_pages(dir: &Path) -> String {
    match fs::metadata(pages_file(dir)) {
        Err(_) => String::new(),
        Ok(meta) if meta.len() > LIMIT => String::new(),
        // Gone between the two calls is another window sweeping the store, which is
        // a paper to read again rather than anything to report.
        Ok(_) => fs::read_to_string(pages_file(dir)).unwrap_or_default(),
    }
}

/// Writes one paper's words down, whole, the way a note is written: a crash mid
/// write leaves what was there rather than half of what is new.
///
/// Nothing to keep takes the folder with it, so a paper whose words were dropped
/// leaves no folder behind.
fn write_pages(dir: &Path, content: &str) -> Result<(), String> {
    if content.is_empty() {
        // Gone already is what was asked for, so none of this is an error.
        let _ = fs::remove_file(pages_file(dir));
        let _ = fs::remove_file(origin_file(dir));
        let _ = fs::remove_dir(dir);
        return Ok(());
    }

    if content.len() > usize::try_from(LIMIT).unwrap_or(usize::MAX) {
        return Err(format!("{} would keep too much", dir.display()));
    }

    fs::create_dir_all(dir).map_err(|error| cannot("create", dir, &error))?;
    write_atomically(&pages_file(dir), content.as_bytes())
}

/// Which papers the store holds words for, and how many bytes each comes to.
///
/// A record whose origin cannot be read is left out: a folder nobody can name a
/// paper for is a folder nobody can ask about, and the window's own sweep is what
/// clears it.
fn listed(root: &Path) -> Vec<PaperFile> {
    let Ok(entries) = fs::read_dir(root) else {
        // Nothing has ever been kept, so there is nothing to list.
        return Vec::new();
    };

    let mut found = Vec::new();

    for dir in entries.flatten().map(|entry| entry.path()) {
        let Ok(path) = fs::read_to_string(origin_file(&dir)) else {
            continue;
        };

        let size = fs::metadata(pages_file(&dir)).map_or(0, |meta| meta.len());
        if size == 0 {
            continue;
        }

        found.push(PaperFile { path, size });
    }

    found
}

/// The words taken out of a PDF, or nothing where none have been. A paper nobody
/// has read is the ordinary case rather than a failure.
#[tauri::command]
pub fn read_paper_text(app: AppHandle, path: String) -> Result<String, String> {
    Ok(read_pages(&paper_dir(&app, &path)?))
}

/// One paper's words written down, or taken away again when the window sends
/// nothing: a paper that has been deleted, renamed or written over has no words
/// worth keeping.
#[tauri::command]
pub fn write_paper_text(app: AppHandle, path: String, content: String) -> Result<(), String> {
    let dir = paper_dir(&app, &path)?;
    write_pages(&dir, &content)?;

    if !content.is_empty() {
        // The paper's own path, so the store can be listed by paper. Not worth
        // failing the write over; the listing simply passes over a record that
        // cannot say which paper it is.
        let _ = fs::write(origin_file(&dir), &path);
    }

    Ok(())
}

/// Every paper the store holds words for.
#[tauri::command]
pub fn list_paper_texts(app: AppHandle) -> Result<Vec<PaperFile>, String> {
    Ok(listed(&papers_dir(&app)?))
}

#[cfg(test)]
mod tests {
    // A test module is its own scope: what it touches is named here rather than
    // borrowed from the module above.
    use super::{listed, origin_file, pages_file, read_pages, write_pages, LIMIT};
    use crate::paths::folder_key;
    use std::fs;
    use std::path::Path;

    #[test]
    fn a_paper_keeps_its_words_and_its_name_in_one_folder() {
        let dir = Path::new("papers").join(folder_key("/Notes/paper.pdf"));

        assert_eq!(pages_file(&dir).file_name().expect("a name"), "pages.json");
        assert_eq!(origin_file(&dir).file_name().expect("a name"), "origin.txt");
        // The folder is a hash rather than the path, so nothing in the store says
        // where anybody keeps their papers.
        assert!(!dir.to_string_lossy().contains("paper.pdf"));
    }

    #[test]
    fn a_paper_nobody_has_read_reads_as_nothing() {
        let dir = tempfile::tempdir().expect("a temp folder");

        assert_eq!(read_pages(&dir.path().join("none")), String::new());
    }

    #[test]
    fn what_is_written_is_what_is_read_back() {
        let dir = tempfile::tempdir().expect("a temp folder");
        let paper = dir.path().join("one");
        let words = r#"{"path":"/Notes/paper.pdf","pages":[[1,"ink"]]}"#;

        write_pages(&paper, words).expect("the write");
        assert_eq!(read_pages(&paper), words);
    }

    #[test]
    fn nothing_left_to_keep_takes_the_folder_with_it() {
        let dir = tempfile::tempdir().expect("a temp folder");
        let paper = dir.path().join("one");

        write_pages(&paper, "{}").expect("the write");
        fs::write(origin_file(&paper), "/Notes/paper.pdf").expect("the origin");

        write_pages(&paper, "").expect("the clearing");
        assert!(!paper.exists());
        // And clearing one that has gone already is not a failure.
        assert_eq!(write_pages(&paper, ""), Ok(()));
    }

    /// The writer is held to the reader's ceiling. Without that, a window could
    /// write words that come back as nothing for ever after.
    #[test]
    fn words_too_many_to_read_back_are_refused_on_the_way_in() {
        let dir = tempfile::tempdir().expect("a temp folder");
        let paper = dir.path().join("one");
        let too_much = "x".repeat(usize::try_from(LIMIT).unwrap_or(usize::MAX) + 1);

        assert!(write_pages(&paper, &too_much).is_err());
        assert_eq!(read_pages(&paper), String::new());
    }

    #[test]
    fn the_store_lists_a_paper_by_the_path_it_came_from() {
        let dir = tempfile::tempdir().expect("a temp folder");
        let one = dir.path().join(folder_key("/Notes/paper.pdf"));
        let two = dir.path().join(folder_key("/Notes/other.pdf"));

        write_pages(&one, "{\"pages\":[]}").expect("the write");
        fs::write(origin_file(&one), "/Notes/paper.pdf").expect("the origin");
        // A record that cannot say which paper it is, which is a half written one.
        write_pages(&two, "{}").expect("the second write");

        let found = listed(dir.path());
        assert_eq!(found.len(), 1);
        assert_eq!(found[0].path, "/Notes/paper.pdf");
        assert_eq!(found[0].size, 12);
    }
}
